### ISSUE TITLE
After failover occurred, commands such as keys() fails due to ConnectionError 

### ISSUE THREAD

**Issue by priancho** [Grokzen/redis-py-cluster#204]
* Library versions:
  * redis (2.10.5)
  * redis-py-cluster (1.3.4)

Hi, 
First, thank you for this great and useful library!

To test Redis cluster master/slave mode, I installed 6 redis instances (3masters and 3slaves) on the same machine with 6 different ports.
Let's say we have the following six nodes.
```
127.0.0.1:37074 master
127.0.0.1:37075 master
127.0.0.1:37076 master
127.0.0.1:37077 slave
127.0.0.1:37078 slave
127.0.0.1:37079 slave
```

Then, I killed one of master nodes.
When I checked with cluster_nodes() method, the change (failover) is applied.
```
all_startup_nodes = [
    {"host":"localhost", "port":37074},
    {"host":"localhost", "port":37075},
    {"host":"localhost", "port":37076},
    {"host":"localhost", "port":37077},
    {"host":"localhost", "port":37078},
    {"host":"localhost", "port":37079}]
conn1 = rediscluster.StrictRedisCluster(startup_nodes=all_startup_nodes)

for node in conn1.cluster_nodes():
    print node["host"], node["port"], node["flags"]
127.0.0.1 37074 ('myself', 'master')
127.0.0.1 37075 ('master', 'fail')
127.0.0.1 37076 ('slave',)
127.0.0.1 37077 ('slave',)
127.0.0.1 37078 ('master',)
127.0.0.1 37079 ('master',)
```

But, as soon as I run some commands such as keys(), it failed as follows.
```
conn1.keys()
---------------------------------------------------------------------------
ConnectionError                           Traceback (most recent call last)
<ipython-input-94-53843ab27ad5> in <module>()
----> 1 conn1.keys()

/opt/anaconda2/lib/python2.7/site-packages/redis/client.pyc in keys(self, pattern)
    934     def keys(self, pattern='*'):
    935         "Returns a list of keys matching ``pattern``"
--> 936         return self.execute_command('KEYS', pattern)
    937
    938     def mget(self, keys, *args):

/opt/anaconda2/lib/python2.7/site-packages/rediscluster/utils.pyc in inner(*args, **kwargs)
     99         for _ in range(0, 3):
    100             try:
--> 101                 return func(*args, **kwargs)
    102             except ClusterDownError:
    103                 # Try again with the new cluster setup. All other errors

/opt/anaconda2/lib/python2.7/site-packages/rediscluster/client.pyc in execute_command(self, *args, **kwargs)
    317         node = self.determine_node(*args, **kwargs)
    318         if node:
--> 319             return self._execute_command_on_nodes(node, *args, **kwargs)
    320
    321         # If set externally we must update it before calling any commands

/opt/anaconda2/lib/python2.7/site-packages/rediscluster/client.pyc in _execute_command_on_nodes(self, nodes, *args, **kwargs)
    408                     raise
    409
--> 410                 connection.send_command(*args)
    411                 res[node["name"]] = self.parse_response(connection, command, **kwargs)
    412             finally:

/opt/anaconda2/lib/python2.7/site-packages/redis/connection.pyc in send_command(self, *args)
    561     def send_command(self, *args):
    562         "Pack and send a command to the Redis server"
--> 563         self.send_packed_command(self.pack_command(*args))
    564
    565     def can_read(self, timeout=0):

/opt/anaconda2/lib/python2.7/site-packages/redis/connection.pyc in send_packed_command(self, command)
    536         "Send an already packed command to the Redis server"
    537         if not self._sock:
--> 538             self.connect()
    539         try:
    540             if isinstance(command, str):

/opt/anaconda2/lib/python2.7/site-packages/redis/connection.pyc in connect(self)
    440         except socket.error:
    441             e = sys.exc_info()[1]
--> 442             raise ConnectionError(self._error_message(e))
    443
    444         self._sock = sock

ConnectionError: Error 111 connecting to 127.0.0.1:37075. Connection refused.
```

While digging the source code, I found that these commands use determine_node("some command") method to figure out Redis servers.
However, the results of this method is old and does not reflect failover.
For example, the following code shows that the killed master node (localhost:37075) is still a master node.
```
for node in list(conn1.determine_node("KEYS")):
    print node
{'host': u'127.0.0.1', 'server_type': 'master', 'port': 37074, 'name': '127.0.0.1:37074'}
{'host': u'127.0.0.1', 'server_type': 'master', 'port': 37075, 'name': '127.0.0.1:37075'}
{'host': u'127.0.0.1', 'server_type': 'slave', 'port': 37076L, 'name': '127.0.0.1:37076'}
{'host': u'127.0.0.1', 'server_type': 'slave', 'port': 37077L, 'name': '127.0.0.1:37077'}
{'host': u'127.0.0.1', 'server_type': 'slave', 'port': 37078L, 'name': '127.0.0.1:37078'}
{'host': u'127.0.0.1', 'server_type': 'master', 'port': 37079, 'name': '127.0.0.1:37079'}
```

I wonder if this is a bug or I need to set some config option.

Best wishes,
Han-Cheol

**Comment by priancho** [comment]
While trying to make it work, I found that NodeManager's reset() method refresh cluster information as follows.

``` 
all_startup_nodes = [
    {"host":"localhost", "port":37074},
    {"host":"localhost", "port":37075},
    {"host":"localhost", "port":37076},
    {"host":"localhost", "port":37077},
    {"host":"localhost", "port":37078},
    {"host":"localhost", "port":37079}]
conn3 = rediscluster.StrictRedisCluster(
    startup_nodes=all_startup_nodes,
    socket_timeout=5)
 
# reset node information
conn3.connection_pool.nodes.reset()
 
for node in list(conn3.determine_node("KEYS")):
    print node["name"],node["server_type"]
127.0.0.1:37079 master
127.0.0.1:37078 master
127.0.0.1:37075 slave
127.0.0.1:37074 master    <-- this was a slave, but promoted when an original master is killed.
127.0.0.1:37076 slave
```

So, we can call reset() method before any command and it will work on refreshed Redis nodes.

But, thinking about that a sequence of these commands (reset() and some command) is not atomic, the command can fail if a Redis cluster node fail again between reset() and command.

Hope to have some comments from other users ;-)


Best wishes,
Han-Cheol

**Comment by Grokzen** [comment]
@priancho Hi

The suggestion that reset() should be called before each command is horrible because then it would query the cluster on every command and slowing down the servers a lot. This will never happen.

How other parts of the code is designed to handle the master failover problem is that you loop over the current view of the cluster and if you have problems to connect like you shown in the first post, then it should raise a connection error exception and that exception should indicate that the `reset()` needs to be used to reset the cluster view to the new reality. My guess is that the `KEYS` method that you tried uses a old code path that do not have this logic implemented properly. You get the exception and that is correct, but it is not handled and that is the bug that needs to be fixed.

Should be fairly easy to track down and fix. Will look at it.

**Comment by jeffwidman** [comment]
>  But it appears that the basic error handling code for this is located in a better state in the method `execute_command` but not in the method `_execute_command_on_nodes`

@Grokzen I'm a little unclear reading your comment... do you consider this to be a legit bug that needs to be cleaned up at some point?

**Comment by Grokzen** [comment]
If i remember correctly, yes that is an major issue that there is 2 different paths through the code where one is the proper one and one is the subpar one that has flaws like the error handling. I do want to join the 2 paths at some point and use a single solution for both types of commands. But that is a major investment that i have hard to put in right now to sort it out.

**Comment by jeffwidman** [comment]
Thanks for clarifying.

**Comment by ofhellsfire** [comment]
It looks like this issue affects us and it is again about **redis** cluster failover.
@Grokzen please let me know if this issue has the same root cause, otherwise it makes more sense to open a new one.

**Issue Description:**
Bunch of exceptions occured (`redis.exceptions.ConnectionError` is one of those) when `scan_iter()` method is called during failover.

## Scenario

**Env**
- Linux OS
- Redis Cluster: 3 master + 3 slave nodes that runs via Docker-Compose locally
- python 3.6.12
- redis-py version: 3.5.3
- redis-py-cluster: 2.1.0

**Steps to reproduce**
- Create Redis Cluster with 6 (3+3) nodes as minimum.
- Start the following script https://github.com/ofhellsfire/redis-cluster-dumb-tests/blob/master/redis-py-cluster/rediscluster_failover_scan_iter_test.py or anything that continuously calls `scan_iter()` method.
- While the script is running, stop of one of masters in the **Redis Cluster** and wait until failover completes.

**Expected result**
Script proceeds without exceptions.

**Actual Result**
Script gets stuck and there are a bunch of exceptions are raised.

**Output**
```
...
Key: 8: 2020-09-15 17:16:54.135722
Key: 13: 2020-09-15 17:16:54.135748
Key: 49: 2020-09-15 17:16:54.135776
Traceback (most recent call last):
  File "rediscluster_failover_scan_iter_test.py", line 25, in <module>
    for key in rc.scan_iter(match='*', count=10):
  File "/home/venv/lib/python3.6/site-packages/rediscluster/client.py", line 969, in scan_iter
    raw_resp = conn.read_response()
  File "/home/venv/lib/python3.6/site-packages/redis/connection.py", line 739, in read_response
    response = self._parser.read_response()
  File "/home/venv/lib/python3.6/site-packages/redis/connection.py", line 324, in read_response
    raw = self._buffer.readline()
  File "/home/venv/lib/python3.6/site-packages/redis/connection.py", line 256, in readline
    self._read_from_socket()
  File "/home/venv/lib/python3.6/site-packages/redis/connection.py", line 201, in _read_from_socket
    raise ConnectionError(SERVER_CLOSED_CONNECTION_ERROR)
redis.exceptions.ConnectionError: Connection closed by server.
Key: 45: 2020-09-15 17:16:54.139949
...
Key: 49: 2020-09-15 17:16:54.140345
Traceback (most recent call last):
  File "/home/venv/lib/python3.6/site-packages/redis/connection.py", line 559, in connect
    sock = self._connect()
  File "/home/venv/lib/python3.6/site-packages/redis/connection.py", line 615, in _connect
    raise err
  File "/home/venv/lib/python3.6/site-packages/redis/connection.py", line 603, in _connect
    sock.connect(socket_address)
OSError: [Errno 113] No route to host

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "rediscluster_failover_scan_iter_test.py", line 25, in <module>
    for key in rc.scan_iter(match='*', count=10):
  File "/home/venv/lib/python3.6/site-packages/rediscluster/client.py", line 967, in scan_iter
    conn.send_command(*pieces)
  File "/home/venv/lib/python3.6/site-packages/redis/connection.py", line 726, in send_command
    check_health=kwargs.get('check_health', True))
  File "/home/venv/lib/python3.6/site-packages/redis/connection.py", line 698, in send_packed_command
    self.connect()
  File "/home/venv/lib/python3.6/site-packages/redis/connection.py", line 563, in connect
    raise ConnectionError(self._error_message(e))
redis.exceptions.ConnectionError: Error 113 connecting to 172.19.0.5:6379. No route to host.
Key: 45: 2020-09-15 17:18:01.548966
...
```

**Comment by Grokzen** [comment]
@ofhellsfire Well kinda yes and no at the same time. The main problem with these lets call them "cluster aware" methods or whatever we shall call them, is that they have some kind of custom cluster implementation where it enables the method to understand and work across multiple slots at the same time. Becuase of that custom implementation, the error logic that exists in the main `execute_command` method do not always apply in the exact same way. For reference see the scan_iter custom implementation here https://github.com/Grokzen/redis-py-cluster/blob/master/rediscluster/client.py#L935

**Comment by Grokzen** [comment]
I do think tho that `SCAN_ITER` is one of the very few custom implementations that actually performs this kind of janky logic and connection handling under the hood so the problem should be very limited in scope really to solve. And it is not the first time that the `SCAN_ITER` method have caused me headache really... `PFMERGE` mig be a close second to this as well.
In 3.0.0 i think i will plan in to acctually split the client into a regular client that do not have these strange custom implementations and one sub class client that has them and the user can choose what level of features to have in it to mirror other clients much closer.

**Comment by Grokzen** [comment]
@ofhellsfire Please take out your specific error that you have above and make a new issue with it so i can track it separate from this and add it into the 3.0.0 release instead.

If i read the root cause of the original issue in this issue then it should be solved now in master branch with the same fix that i applied to your other issue @ofhellsfire as the node selection and execution path is now covered in both scenarios with the try/except handling now so the connection error should no longer happen the same way as it does during failover.

**Comment by Grokzen** [comment]
@priancho If possible please try the latest commit in master to verify that the issue is resolved now. I will close this issue in a while if no update is given so you know.

**Comment by ofhellsfire** [comment]
> @ofhellsfire Please take out your specific error that you have above and make a new issue with it so i can track it separate from this and add it into the 3.0.0 release instead.

@Grokzen Done!

**Comment by Grokzen** [comment]
Will close this issue as resolved and if the problem still remains in 2.1.0 then please open up a new issue and it can be solved in there.
